safe_control_gym/controllers/mpc/mpc_acados.py

```diff
@@ -0,0 +1,398 @@
+              'he': set_acados_constraint_bound(he_expr, 'lb'), }
+
+        # make sure all the ub and lb are 1D numpy arrays
+        # (see: https://discourse.acados.org/t/infeasible-qps-when-using-nonlinear-casadi-constraint-expressions/1595/5?u=mxche)
+        for key in ub.keys():
+            ub[key] = ub[key].flatten() if ub[key].ndim != 1 else ub[key]
+            lb[key] = lb[key].flatten() if lb[key].ndim != 1 else lb[key]
+        # check ub and lb dimensions
+        for key in ub.keys():
+            assert ub[key].ndim == 1, f'ub[{key}] is not 1D numpy array'
+            assert lb[key].ndim == 1, f'lb[{key}] is not 1D numpy array'
+        assert ub['h'].shape == lb['h'].shape, 'h_ub and h_lb have different shapes'
+
+        # pass the constraints to the ocp object
+        ocp.model.con_h_expr_0, ocp.model.con_h_expr, ocp.model.con_h_expr_e = \
+            h0_expr, h_expr, he_expr
+        ocp.dims.nh_0, ocp.dims.nh, ocp.dims.nh_e = \
+            h0_expr.shape[0], h_expr.shape[0], he_expr.shape[0]
+        # assign constraints upper and lower bounds
+        ocp.constraints.uh_0 = ub['h0']
+        ocp.constraints.lh_0 = lb['h0']
+        ocp.constraints.uh = ub['h']
+        ocp.constraints.lh = lb['h']
+        ocp.constraints.uh_e = ub['he']
+        ocp.constraints.lh_e = lb['he']
+
+        return ocp
+
+    @timing
+    def select_action(self,
+                      obs,
+                      info=None
+                      ):
+        '''Solves nonlinear mpc problem to get next action.
+
+        Args:
+            obs (ndarray): Current state/observation.
+            info (dict): Current info
+
+        Returns:
+            action (ndarray): Input/action to the task/env.
+        '''
+        nx, nu = self.model.nx, self.model.nu
+        # set initial condition (0-th state)
+        self.acados_ocp_solver.set(0, 'lbx', obs)
+        self.acados_ocp_solver.set(0, 'ubx', obs)
+
+        # warm-starting solver
+        # NOTE: only for ipopt warm-starting; since acados
+        # has a built-in warm-starting mechanism.
+        if self.warmstart:
+            if self.x_guess is None or self.u_guess is None:
+                # compute initial guess with IPOPT
+                self.compute_initial_guess(obs)
+            for idx in range(self.T + 1):
+                init_x = self.x_guess[:, idx]
+                self.acados_ocp_solver.set(idx, 'x', init_x)
+            for idx in range(self.T):
+                if nu == 1:
+                    init_u = np.array([self.u_guess[idx]])
+                else:
+                    init_u = self.u_guess[:, idx]
+                self.acados_ocp_solver.set(idx, 'u', init_u)
+
+        # set reference for the control horizon
+        goal_states = self.get_references()
+        if self.mode == 'tracking':
+            self.traj_step += 1
+
+        y_ref = np.concatenate((goal_states[:, :-1], np.zeros((nu, self.T))))
```

Comment:
should this be `np.zeros((nu, self.T))` or `self.U_EQ`?

Reply:
Thank you for pointing this out. I agree `self.U_EQ` could make more sense and I will fix this.